### DISCUSSION TITLE
Why are the two sets of shap values returned in binary classification tasks not the inverse of one another?

### DISCUSSION THREAD

**Discussion by jaxondk** [shap/shap#1502] [Q&A]
Hello all!

I'm certain that I'm revealing my very limited understanding with this post, but I am confused as to why `DeepExplainer.shap_values` returns a list (call it `shap_values`) of NxF 2d arrays of shap values for a binary classification task, where N is the number of explanatory samples and F is the number of features, instead of just one NxF 2d array. Similarly, I'm confused why my explainer object as two expected values instead of just one. This brings me to two main questions.

### Why?
I know that it's returning one NxF 2d array of shap values for class 0 and also one for class 1 (and one expected value for each class as well), but I don't understand why. My limited understanding of the theory behind SHAP values is that we are using some kind of approximation algorithm for the power set of possible feature combinations, so that we can learn each feature's marginal contribution in reducing the model's loss. This is done by seeing the deviation from the expected value of the model caused by the addition of a feature, and doing that for all the features. Why does the model have a different expected value for class 0 vs. class 1?

And further, when we say that `shap_values[0]` are the shap values for class 0, or `expected_value[0]` is the expected value for class 0, what do we mean? It's not the shap values for samples that are labeled class 0 (nor predicted class 0), since `shap_values[0]` has N rows and not just the subset of rows that are labeled class 0. What is it that we are changing when we calculate shap values for class 0 as opposed to class 1?

### How to use both?
How do I use the two sets of shap values in practice? Generating plots using `shap_values[0]` vs `shap_values[1]` results in very different results and conclusions about the data. `shap.summary_plot(..., plot_type='bar')` can take in my entire `shap_values` array and stack class 0 and class 1, but most shap plotting functions require me to pick one (`shap_values[0]` or `shap_values[1]`). I am mostly using global plots - I assume for local plots where we're only looking at one sample, I should use the shap values that correspond to the predicted class for that point. 

Most examples I've seen do not address this question and are only passing `shap_values`, which does not work for me (IE errors are thrown) except for the summary bar plot. The few examples that don't do this seem to use `shap_values[1]`, vaguely mentioning that since we usually care more about the positive class that's what we should look at. But if we're trying to understand why the model is predicting the way it's predicting, don't we care about both classes?

Should I be using just one of the two shap value arrays for drawing conclusions on the data or both? If it depends on what kinds of questions I'm trying to answer, please give some examples of when you would use both vs. when you would use class 0 only vs. when you would use class 1 only.

### Additional Context
Despite using a DL model, my data is tabular, so if anything is specific to image or text classification it does not apply to me

I've seen [this github issue](https://github.com/slundberg/shap/issues/526#issue-428404125) that suggests that the shap values of the negative class are simply the negation of the shap values of the positive class (though this discussion is about LightGBM and tree explainer). Comparing `shap_values[0]` with `shap_values[1]`, this is not the case for me, as mine aren't just the negation of the other

Any insight is appreciated :pray:

**Comment by jaxondk** [comment]
I'm replying to my own comment with some things I've learned. Please feel free to correct me.

### Why 2 sets of shap values
First, a binary classification model always outputs a 2xN array of prediction values, where N is the size of your test set. The first column of prediction values are (at least for my model) the logits for class 0, the second column are the logits for class 1. This is fairly obvious. The expected value is just the mean of these, which is a 2x1. This answers why there are 2 expected values. The shap values are then calculated using each one of these expected values by looking at the model output for a given row, finding the difference between this row's model output and the expected value, and then calculating the marginal contribution of each feature that caused the difference. Thus the 2 sets of shap values.

### Updated question: Why are the two sets of shap values returned in binary classification tasks not the inverse of one another?
I would expect that the beeswarm summary plots for class 0 would then just be the reflection of the beeswarm summary plots for class 1, but this is not the case for every feature. Why? How should they relate to one another then? Similarly, I would expect the stacked summary bar plots (where the classes are stacked) to generate the same rankings of features as the summary bar plots for a single class. They do not, and I do not understand why

**Comment by letsdecker** [comment]
Do you have any news? I have the same doubt.
